### ISSUE TITLE
Mark unrecognized characters in the output result

### ISSUE THREAD

**Issue by ssmelov** [openalpr/openalpr#265]
For example, I have a license plate 'ABC123'.
If OpenALPR fails to recognize '1' character (because of noise or any other reason) then it reports: 'ABC23' with quite high confidence level (about 75-85%).

Wouldn't it be better to produce the following output result: ABC?23 and set lower confidence level to mark that we know that there is a character, but we failed to recognize it(or not sure that this character is correct)?

**Comment by matthill** [comment]
Yes, perhaps.  Many times there's a decal, or graphic on the plate that we want to ignore, since it's not a letter.  

I wonder if there's a good way to disambiguate.  Perhaps if it's narrowed down to 1 or 2 possible characters, and the confidence isn't *too* low.

**Comment by alandtse** [comment]
If you want to see the effect with your data set, a quick edit to postprocess.cpp:360 to remove the 
if statement check for "(letter.letter != SKIP_CHAR)" will give you a taste of the results.  My CA data sets don't have decal characters so it's not blowing anything up directly, but your mileage may vary.  I think lowering the total confidence score may have to occur after you've selected that permutation as the best though.  Otherwise, that particular SKIP_CHAR may always lose to the alternatives and it'd never surface as a result.

Matt: Looking at that section of code in general without that edit, is there a bug in that the line for totaling the score is still adding the SKIP_CHAR score when the skip is not being added? Admittedly I just saw it, so maybe you wanted that behavior.

        possibility.totalscore = possibility.totalscore + letter.totalscore;

**Comment by ssmelov** [comment]
Hey I started to work on this one, but it creates another problem which is related to the thresholding.
Here is a quick example: I have a license plate 123ABC and there is gray spot between 123 and ABC. Thresholding thinks that this gray spot is a character and sends it to Tesseract for recognition. All the recognitions for this "gray spot" return low confidence levels and this character should be skept, but it will return question mark if we implement the option to display unrecognized characters: 123?ABC. Which is completely wrong and breaks most of my tests.

After I had a look on the thresholds the OpenALPR produces it looks like they are not perfect and include  a lot of noise which causes some characters to be recognized incorrectly or some gray/dirt areas on the license plate to be recognized as characters.

I would like to play around a bit with thresholding to see if we can improve it and reduce the noise.
Here is one of the thresholding algorithms I would like to try: http://www.ijarcce.com/upload/2015/october-15/IJARCCE%20124.pdf

**Comment by matthill** [comment]
That is an interesting thresholding approach.  It could be better than what's currently being used.  It may still pick up the middle decal, though.  The thresholding is really just trying to pick out dark from light (or vice-versa).  So it makes sense for it to include the decal since there is a contrast.  Typically OCR will filter that stuff out but not always.

I've tried color filtering in the past (i.e., get the median Hue for the character locations and filter all the other stuff out).  But this did not work well on lighter/darker plate areas (the hue is mostly lost).  

An approach that I think will work is a CNN classifier.  The MNIST classifier (http://caffe.berkeleyvision.org/gathered/examples/mnist.html) is geared towards recognizing handwritten digits.  Given a sufficiently large database of good versus bad black and white character crops, you could train the classifier to accept or reject characters before they're sent to OCR.  This would accept the 6 characters, but drop the decal.
